source/solvers/mf_navier_stokes_operators.cc

```diff
@@ -0,0 +1,810 @@
+  for (unsigned int l = 0; l < nlevels; l++)
+    refinement_edge_indices[l] =
+      IndexSet(matrix_free.get_dof_handler().n_dofs(l));
+
+  MGTools::extract_inner_interface_dofs(matrix_free.get_dof_handler(),
+                                        refinement_edge_indices);
+  return refinement_edge_indices[level];
+}
+
+template class NavierStokesOperatorBase<2, double>;
+template class NavierStokesOperatorBase<3, double>;
+
+/**
+ * @brief This function performs a cell integral, i.e., the jacobian of the discretization
+ * of the Navier-Stokes equations with SUPG PSPG stabilization in a cell batch.
+ * The equations are given as follows:
+ * (q,∇δu) + (v,(u·∇)δu) + (v,(δu·∇)u) - (∇·v,δp) + ν(∇v,∇δu) (Weak form
+ * jacobian)
+ * + ((u·∇)δu + (δu·∇)u + ∇δp - ν∆δu)τ·∇q (PSPG Jacobian)
+ * + ((u·∇)δu + (δu·∇)u + ∇δp - ν∆δu)τu·∇v (SUPG Part 1)
+ * + ((u·∇)u + ∇p - ν∆u - f )τδu·∇v (SUPG Part 2)
+ */
+template <int dim, typename number>
+NavierStokesSUPGPSPGOperator<dim, number>::NavierStokesSUPGPSPGOperator()
+{}
+
+template <int dim, typename number>
+void
+NavierStokesSUPGPSPGOperator<dim, number>::do_cell_integral_local(
+  FECellIntegrator &integrator) const
+{
+  integrator.evaluate(EvaluationFlags::values | EvaluationFlags::gradients |
+                      EvaluationFlags::hessians);
+
+  const unsigned int cell = integrator.get_current_cell_index();
+
+  auto h = integrator.read_cell_data(this->get_element_size());
+
+  for (unsigned int q = 0; q < integrator.n_q_points; ++q)
+    {
+      // Evaluate source term function
+      Tensor<1, dim + 1, VectorizedArray<double>> source_value;
+      Point<dim, VectorizedArray<double>>         point_batch =
+        integrator.quadrature_point(q);
+      source_value =
+        evaluate_function<dim, double, dim + 1>(*(this->forcing_function),
+                                                point_batch);
+
+      // Gather the original value/gradient
+      typename FECellIntegrator::value_type    value = integrator.get_value(q);
+      typename FECellIntegrator::gradient_type gradient =
+        integrator.get_gradient(q);
+      typename FECellIntegrator::gradient_type hessian_diagonal =
+        integrator.get_hessian_diagonal(q);
+
+      // Result value/gradient we will use
+      typename FECellIntegrator::value_type    value_result;
+      typename FECellIntegrator::gradient_type gradient_result;
+
+      // Gather previous values of the velocity and the pressure
+      auto previous_values   = this->nonlinear_previous_values(cell, q);
+      auto previous_gradient = this->nonlinear_previous_gradient(cell, q);
+      auto previous_hessian_diagonal =
+        this->nonlinear_previous_hessian_diagonal(cell, q);
+
+      // Calculate tau
+      VectorizedArray<number> u_mag = VectorizedArray<number>(0.0);
+      VectorizedArray<number> tau   = VectorizedArray<number>(0.0);
+      for (unsigned int k = 0; k < dim; ++k)
+        {
+          u_mag += previous_values[k];
+        }
```

Comment:
```suggestion
      for (unsigned int k = 0; k < dim; ++k)
          u_mag += previous_values[k];
```
one line for loop are cleaning I think in this case